target/linux/mediatek/patches-6.6/902-arm-add-cmdline-replace.patch

```diff
@@ -0,0 +1,71 @@
+--- a/arch/arm64/Kconfig
++++ b/arch/arm64/Kconfig
+@@ -2316,6 +2316,15 @@ config CMDLINE_OVERRIDE
+ 	  devices. This setting enables using "chosen/cmdline-override" as the
+ 	  cmdline if it exists in the device tree.
+ 
++config CMDLINE_REPLACE
++       bool "Replace specific cmdline from bootloader"
++	default n
++       help
++         Some bootloaders pass specific bootargs to the linux kernel that need
++         to additional behaviors for proper operation. This setting enables
++         replacing bootargs from "chosen/cmdline-replace" at early boot if it
++         exists in the device tree.
++
+ config EFI_STUB
+ 	bool
+ 
+--- a/drivers/of/fdt.c
++++ b/drivers/of/fdt.c
+@@ -1157,6 +1157,11 @@ int __init early_init_dt_scan_chosen(cha
+ 	const void *rng_seed;
+ 	const void *fdt = initial_boot_params;
+ 
++#ifdef CONFIG_CMDLINE_REPLACE
++	char *delim;
++	char *str;
++#endif
++
+ 	node = fdt_path_offset(fdt, "/chosen");
+ 	if (node < 0)
+ 		node = fdt_path_offset(fdt, "/chosen@0");
+@@ -1200,6 +1205,38 @@ int __init early_init_dt_scan_chosen(cha
+ 		strlcpy(cmdline, p, min((int)l, COMMAND_LINE_SIZE));
+ #endif
+ 
++    /* CONFIG_CMDLINE_REPLACE is used to replace a kernel argument at early-
++     * boot with a different device tree option in chosen/bootargs-replace.
++     * This is helpful for boards where u-boot sets bootargs, and is need
++     * to be identified and replaced.
++     */
++#ifdef CONFIG_CMDLINE_REPLACE
++	p = of_get_flat_dt_prop(node, "bootargs-replace", &l);
++	if (p != NULL && l > 0) {
++		if ((delim = strstr(p, "|")) != NULL) {
++			*delim = '\0';
++
++			/* bound check and is replace available */
++			if (!(p + l < delim + 1) &&
++				(str = strstr(cmdline, p)) != NULL) {
++				if (strlen(p) != strlen(delim + 1))
++					memmove(str + strlen(delim + 1), str + strlen(p),
++						min_t(int, strlen(str + strlen(p)) + 1,
++						COMMAND_LINE_SIZE));
```

Comment:
The full `cmdline` is max `COMMAND_LINE_SIZE long. This is only the tail.

Reply:
I think it would be better to extract some of this into an own function to make it easier understandable.
The kernel already has `strreplace`, but it does not support changing size.

Reply:
yes, I think seperate for create own function is make it understandable too.

I will modifying it to seperate and limit to `COMMAND_LINE_SIZE` when copying the full `cmdline` at the end
later.

Reply:
Extracted as a function named `string_replace`.